### PR TITLE
[fix][io] KCA sink: handle null values with KeyValue<Avro,Avro> schema

### PR DESCRIPTION
**Pull request by nicoloboschi**
### Motivation

When using a KCA based sink connector with a topic with schema KeyValue<Avro, Avro>, if a message has a null value, the connector throws:
```
023-03-15T08:00:22,139+0000 [public/default/sink] ERROR org.apache.pulsar.io.kafka.connect.KafkaConnectSink - Error sending the record SinkRecord(sourceRecord=PulsarRecord(topicName=Optional[persistent://public/default/data-baselines.keyvalue_short_tests], partition=0, message=Optional[org.apache.pulsar.client.impl.MessageImpl@2ddf2db4], schema=KeyValueSchema(SEPARATED,org.apache.pulsar.client.impl.schema.generic.GenericAvroSchema@33df2fa0,org.apache.pulsar.client.impl.schema.generic.GenericAvroSchema@359ba6c9), failFunction=org.apache.pulsar.functions.source.PulsarSource$$Lambda$264/0x00000008404bb440@227fd2ce, ackFunction=org.apache.pulsar.functions.source.PulsarSource$$Lambda$263/0x00000008404bb040@7619e9b8), value=(key = "org.apache.pulsar.client.impl.schema.generic.GenericAvroRecord@3c717af4", value = "null"))
org.apache.kafka.connect.errors.DataException: Invalid null value for required STRUCT field
	at org.apache.pulsar.io.kafka.connect.schema.KafkaConnectData.defaultOrThrow(KafkaConnectData.java:385) ~[pulsar-io-kafka-connect-adaptor-2.10.3.5.jar:2.10.3.5]
	at org.apache.pulsar.io.kafka.connect.schema.KafkaConnectData.getKafkaConnectData(KafkaConnectData.java:66) ~[pulsar-io-kafka-connect-adaptor-2.10.3.5.jar:2.10.3.5]
	at org.apache.pulsar.io.kafka.connect.KafkaConnectSink.toSinkRecord(KafkaConnectSink.java:442) ~[pulsar-io-kafka-connect-adaptor-2.10.3.5.jar:2.10.3.5]
	at org.apache.pulsar.io.kafka.connect.KafkaConnectSink.write(KafkaConnectSink.java:125) ~[pulsar-io-kafka-connect-adaptor-2.10.3.5.jar:2.10.3.5]
	at org.apache.pulsar.functions.instance.JavaInstanceRunnable.sendOutputMessage(JavaInstanceRunnable.java:424) ~[?:?]
	at org.apache.pulsar.functions.instance.JavaInstanceRunnable.handleResult(JavaInstanceRunnable.java:394) ~[?:?]
	at org.apache.pulsar.functions.instance.JavaInstanceRunnable.run(JavaInstanceRunnable.java:335) ~[?:?]
	at java.lang.Thread.run(Thread.java:829) ~[?:?]
```
### Modifications

Handle null value conversion to the kafka value

### Verifying this change

- [x] Make sure that the change passes the CI checks.

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete`

### CODE REVIEW

**Review by eolivelli — APPROVED** [review]
Lgtm
Nice catch

**Comment by eolivelli** [comment]
actually in Pulsar a KeyValue<?,?> schema allows null for both the key and the value.

@dlg99 in Pulsar it is not allowed to nest KeyValue schemas. KeyValue schema is a (bad) trick to provide a Schema to the key of the message.

any usage of nested KeyValue schemas is not supported (we may issue a warning or throw an error in Pulsar 3.0) and it cannot work properly

so this schema doesn't work 
`Schema.KeyValue(Schema.INT32, Schema.KeyValue(Schema.INT32, Schema.KeyValue(Schema.INT32, schema)))`

**Comment by dlg99** [comment]
> actually in Pulsar a KeyValue schema allows null for both the key and the value.
> 
> @dlg99 in Pulsar it is not allowed to nest KeyValue schemas. KeyValue schema is a (bad) trick to provide a Schema to the key of the message.
> 
> any usage of nested KeyValue schemas is not supported (we may issue a warning or throw an error in Pulsar 3.0) and it cannot work properly
> 
> so this schema doesn't work `Schema.KeyValue(Schema.INT32, Schema.KeyValue(Schema.INT32, Schema.KeyValue(Schema.INT32, schema)))`

that's fine, nest any other pulsar schema.

The problem would loss of optional flag in nested schema in translation form kafka schema to pulsar (source -> topic -> trabsform -> sink chain)

File: pulsar-io/kafka-connect-adaptor/src/test/java/org/apache/pulsar/io/kafka/connect/KafkaConnectSinkTest.java [hunk 1]
```diff
@@ -690,6 +694,20 @@ public void schemaKeyValueSchemaTest() throws Exception {
...
+        builder.field("test").type(SchemaType.STRING);
+        GenericSchema<GenericRecord> schema = GenericAvroSchema.of(builder.build(SchemaType.AVRO));
+        KeyValue<Integer, String> kv = new KeyValue<>(11, null);
+        SinkRecord sinkRecord = recordSchemaTest(kv, Schema.KeyValue(Schema.INT32, schema), 11,
+                "INT32", null, "STRUCT");
+        Assert.assertNull(sinkRecord.value());
+        int key = (int) sinkRecord.key();
+        Assert.assertEquals(key, 11);
+    }
+
```
Comment (dlg99) [inline pulsar-io/kafka-connect-adaptor/src/test/java/org/apache/pulsar/io/kafka/connect/KafkaConnectSinkTest.java:710]:
```suggestion

    @Test
    public void schemaKeyValueSchemaNullValueNoUnwrapTest() throws Exception {
        props.put("unwrapKeyValueIfAvailable", "false");
        JSONSchema<PulsarSchemaToKafkaSchemaTest.StructWithAnnotations> jsonSchema = JSONSchema
                .of(SchemaDefinition.<PulsarSchemaToKafkaSchemaTest.StructWithAnnotations>builder()
                        .withPojo(PulsarSchemaToKafkaSchemaTest.StructWithAnnotations.class)
                        .withAlwaysAllowNull(true)
                        .build());

        KeyValue<Integer, String> kv = new KeyValue<>(11, null);

        Map<String, Object> expected = new HashMap();
        expected.put("11", null);
        SinkRecord sinkRecord = recordSchemaTest(kv, Schema.KeyValue(Schema.INT32, jsonSchema), "key",
                "STRING", expected, "MAP");
        Assert.assertNull(((Map)sinkRecord.value()).get(11));
        String key =(String)sinkRecord.key();
        Assert.assertEquals(key, "key");
    }

    @Test
    public void schemaKeyValueSchemaNullValueNoUnwrapTestAvro() throws Exception {
        props.put("unwrapKeyValueIfAvailable", "false");
        RecordSchemaBuilder builder = SchemaBuilder
                .record("test");
        builder.property("op", "test");
        builder.field("test").type(SchemaType.STRING);
        GenericSchema<GenericRecord> schema = GenericAvroSchema.of(builder.build(SchemaType.AVRO));
        KeyValue<Integer, String> kv = new KeyValue<>(11, null);

        Map<String, Object> expected = new HashMap();
        expected.put("11", null);

        SinkRecord sinkRecord = recordSchemaTest(kv, Schema.KeyValue(Schema.INT32, schema), "key",
                "STRING", expected, "MAP");
        Assert.assertNull(((Map)sinkRecord.value()).get(11));
        String key =(String)sinkRecord.key();
        Assert.assertEquals(key, "key");
    }

```

File: pulsar-io/kafka-connect-adaptor/src/main/java/org/apache/pulsar/io/kafka/connect/schema/PulsarSchemaToKafkaSchema.java [hunk 2]
```diff
@@ -80,6 +153,11 @@ private static org.apache.avro.Schema parseAvroSchema(String schemaJson) {
         return parser.parse(schemaJson);
     }
 
+    public static Schema getOptionalKafkaConnectSchema(org.apache.pulsar.client.api.Schema pulsarSchema) {
+        Schema s = getKafkaConnectSchema(pulsarSchema);
+        return new OptionalForcingSchema(s);
```
Comment (eolivelli) [inline pulsar-io/kafka-connect-adaptor/src/main/java/org/apache/pulsar/io/kafka/connect/schema/PulsarSchemaToKafkaSchema.java:158]: we are creating a new object for each record.
maybe we could save this allocation by pushing down OptionalForcingSchema into the cache.
what do you think ?

Comment (dlg99) [reply pulsar-io/kafka-connect-adaptor/src/main/java/org/apache/pulsar/io/kafka/connect/schema/PulsarSchemaToKafkaSchema.java:158]: the only place it is used is within cache's get() call:

```
            return schemaCache.get(pulsarSchema.getSchemaInfo().getSchema(), () -> {
                if (pulsarSchema.getSchemaInfo().getType() == SchemaType.KEY_VALUE) {
                    KeyValueSchema kvSchema = (KeyValueSchema) pulsarSchema;
                    return SchemaBuilder.map(getKafkaConnectSchema(kvSchema.getKeySchema()),
                                             getKafkaConnectSchema(kvSchema.getValueSchema()))
                                    getOptionalKafkaConnectSchema(kvSchema.getValueSchema()))
                                .build();
                }
```

**Review by eolivelli — APPROVED** [review]
LGTM

**Review by dlg99 — APPROVED** [review]

**Pull request merged**